src/main/java/org/springframework/samples/petclinic/rest/importcsv/ImportCSV.java

```diff
@@ -33,107 +40,153 @@
         produces = "application/json")
     public ResponseEntity<List<Pet>> importPets(@RequestBody String csv) {
 
-        int i = 0;
         List<Pet> pets = new LinkedList<Pet>();
         Pet pet;
 
-        do {
-            pet = new Pet();
+        List<String> splitCSV = segment_line_to_entries(csv);
+        pet = convert_string_list_to_pet( splitCSV );
+        // DON'T FORGET TO ADD ERROR HANDLING FOR pet.setOwner:
+        // return new ResponseEntity<List<Pet>>(headers, HttpStatus.BAD_REQUEST);
 
-            String field = "";
-            while (i < csv.length() && csv.charAt(i) != ';') {
-                field += csv.charAt(i++);
-            }
-            i++;
+        if (splitCSV.get(4).toLowerCase().equals("add")) {
+            clinicService.savePet(pet);
+        } else {
+            removePetFromOwner(pet);
+        }
 
-            pet.setName(field);
+        pets.add(pet);
 
-            field = "";
-            while (i < csv.length() && csv.charAt(i) != ';') {
-                field += csv.charAt(i++);
-            }
-            i++;
-
-            try {
-                pet.setBirthDate((new SimpleDateFormat("yyyy-MM-dd")).parse(field));
-            } catch (ParseException e) {
-                HttpHeaders headers = new HttpHeaders();
-                headers.add("errors", "date " + field + " not valid");
-                return new ResponseEntity<List<Pet>>(headers, HttpStatus.BAD_REQUEST);
-            }
+        return new ResponseEntity<List<Pet>>(pets, HttpStatus.OK);
+    }
 
-            field = "";
-            while (i < csv.length() && csv.charAt(i) != ';') {
-                field += csv.charAt(i++);
-            }
-            i++;
-
-            if (pet != null) {
-                ArrayList<PetType> ts = (ArrayList<PetType>) clinicService.findPetTypes();
-                for (int j = 0; j < ts.size(); j++) {
-                    if (ts.get(j).getName().toLowerCase().equals(field)) {
-                        pet.setType(ts.get(j));
-                        break;
-                    }
-                }
-            }
+    private List<String> convert_csv_to_lines(String csv)
```

Comment:
I like your idea to create function `convert_csv_to_lines` and `segment_line_to_entries` :)
Try to use one consistent naming notation (e.g., camel case)